### PR TITLE
persist: split seal into physical and logical components.

### PR DESCRIPTION
**Pull request by ruchirK**

### CODE REVIEW

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:383]: seal being a verb and a noun makes this line confusing. Let's rename the one this is calling to current_seal or get_seal or something

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:394]: Not this change, but I think we actually want an antichain of the minimal element for the since here. since we don't compact in future, a new batch always has the full fidelity of the original data

File: src/persist/src/indexed/mod.rs
Comment (danhhz) [inline src/persist/src/indexed/mod.rs:486]: I think we actually want to call this when we've successfully moved the data into trace. That way, a user that has received a seal knows that it can use any indexed access we add for trace data. The way this is now, some of that data might still be in future when the listener is notified

**Review by danhhz — APPROVED** [review]
LGTM modulo the comment about where listen_fn is getting called

File: src/persist/src/indexed/trace.rs
Comment (ruchirK) [reply src/persist/src/indexed/trace.rs:124]: This is a great question!

Added this response as a comment

```
122     /// While `self.seal` tracks the frontier of times that have been logically been
123     /// closed and are eligible to be moved into the trace, `self.ts_upper()` tracks
124     /// the frontier of times that have actually been physically moved into the trace.
125     /// `self.seal()` is required to manage invariants between commands (e.g. a seal request
126     /// has to be at a time in advance of prior seal requests) whereas `self.ts_upper()`
127     /// is required to manage physical reads and writes to the trace (e.g. to determine
128     /// which times may be added that are not already present.
```

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:486]: I'm happy to make that change but then the behavior would be different between writes and seals because for writes we currently notify listeners as soon as the data lands in buffer. If its the same to you, I would vote that we leave it as is because the other pr which pushes all user requests into a command queue changes things quite a bit and makes it easier to notify listeners when updates get moved into trace.

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:394]: done

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs:383]: done

File: src/persist/src/indexed/mod.rs
Comment (ruchirK) [reply src/persist/src/indexed/mod.rs]: nvm! changed my mind and added it as a separate commit because its basically a prereq to having the other pr work correctly :D

**Comment by ruchirK** [comment]
TFTR! merging on green -- adding the change to make listen events happen when data is moved into the trace required me to introduce an additional write to make sure the data correctly got placed in the trace. I'll rectify that in the next pr which changes `step` to run in the background by adding an explicit test only command for "move data into future/trace"

File: src/persist/src/indexed/trace.rs
Comment (danhhz) [reply src/persist/src/indexed/trace.rs:131]: Makes sense! Great comment :clap:. And yeah, any time I ask a question like this in code review, it's probably a good prompt to add a comment in case someone else later has the same question. (This is something I try to do personally)

**Review by danhhz — COMMENTED** [review]
> I'm happy to make that change but then the behavior would be different between writes and seals because for writes we currently notify listeners as soon as the data lands in buffer.

Looks like github has gc'd my ability to actually respond to this thread. FWIW, I think this response is a false equivalence. What happens internally isn't as important as what the user visible effects are, and in this case, the correct time IMO for a listener to be notified of a seal is when they would be guaranteed to see all the sealed data in the trace

**Pull request merged**
